### PR TITLE
WD-8528 - feat: Display errors in Actions Panel

### PR DESCRIPTION
**Pull request by vladimir-cucu**
## Done

- Display inline error when trying to get actions for application.
- Display error in toast when trying to execute action on units.

## TODO
- Write new tests and fix old tests.
- Remove unused `EXECUTE_ACTION_ERROR` label.

## Details

- https://warthogs.atlassian.net/browse/WD-8528

### CODE REVIEW

File: src/panels/ActionsPanel/ActionsPanel.tsx [hunk 1]
```diff
@@ -213,9 +224,15 @@ export default function ActionsPanel(): JSX.Element {
...
                 setConfirmType(null);
-                executeAction().catch((error) =>
-                  console.error(Label.EXECUTE_ACTION_ERROR, error),
-                );
+                executeAction().catch((error) => {
+                  const errorMessage = `Unable to execute ${selectedAction} action on units: ${generateSelectedUnitList()}.`;
+                  reactHotToast.custom((t) => (
+                    <ToastCard toastInstance={t} type="negative">
+                      {errorMessage}
+                    </ToastCard>
```
[the remark this answers is not in the document]
Comment (vladimir-cucu) [reply src/panels/ActionsPanel/ActionsPanel.tsx]: I've changed it to inline error as discussed in the morning.

File: src/panels/ActionsPanel/ActionsPanel.tsx [hunk 2]
```diff
@@ -211,12 +244,21 @@ export default function ActionsPanel(): JSX.Element {
               cancelButtonLabel={Label.CANCEL_BUTTON}
               confirmButtonLabel={Label.CONFIRM_BUTTON}
               confirmButtonAppearance="positive"
-              onConfirm={() => {
+              onConfirm={(event: React.MouseEvent<HTMLElement, MouseEvent>) => {
+                event.nativeEvent.stopImmediatePropagation();
```
Comment (vladimir-cucu) [inline src/panels/ActionsPanel/ActionsPanel.tsx:248]: This line of code (and maybe similar lines in other places) might become redundant in the case [this gets implemented](https://github.com/canonical/react-components/issues/1032).

File: src/panels/ActionsPanel/ActionsPanel.tsx [hunk 3]
```diff
@@ -119,9 +145,16 @@ export default function ActionsPanel(): JSX.Element {
             setActionData(actions.results[0].actions);
           }
           setFetchingActionData(false);
+          setInlineErrors((prevInlineErrors) => [null, prevInlineErrors[1]]);
```
Comment (huwshimi) [inline src/panels/ActionsPanel/ActionsPanel.tsx:148]: This case (and the other setInlineErrors() calls below) would lead to bugs if the total number of errors in the tuple changed (e.g. if a third error was added it would get lost during this call). Instead we should only update the slot that matches the error we're concerned with.

If you're concerned about mutating state then you could do something like:
```
errors = [... prevInlineErrors];
errors[0] = null;
```

File: src/panels/ActionsPanel/ActionsPanel.tsx (hunk 2 above)
Comment (huwshimi) [reply src/panels/ActionsPanel/ActionsPanel.tsx:248]: Thanks for filing that issue. I think it would be good to leave a comment here about why this `stopImmediatePropagation` is necessary and probably link to the issue you filed as well.

File: src/panels/ActionsPanel/ActionsPanel.tsx [hunk 4]
```diff
@@ -211,12 +244,21 @@ export default function ActionsPanel(): JSX.Element {
...
               confirmButtonAppearance="positive"
-              onConfirm={() => {
+              onConfirm={(event: React.MouseEvent<HTMLElement, MouseEvent>) => {
+                event.nativeEvent.stopImmediatePropagation();
                 setConfirmType(null);
-                executeAction().catch((error) =>
-                  console.error(Label.EXECUTE_ACTION_ERROR, error),
-                );
-                handleRemovePanelQueryParams();
+                executeAction()
```
Comment (huwshimi) [inline src/panels/ActionsPanel/ActionsPanel.tsx:250]: I think this should clear the errors here in case the user submits the form again. Also, depending on how long this action takes there could be a pause before the panel closes. You could show a loading state in the button. You'd need a new state (e.g. "const [saving, setSaving] = useState(false);".

```suggestion
                setSaving(true);
                executeAction()
```

You'll also need to update the button on line 287 to be `<ActionButton` and then give it the props `disabled={saving} loading={saving}`.

File: src/panels/ActionsPanel/ActionsPanel.tsx [hunk 5]
```diff
@@ -211,12 +244,21 @@ export default function ActionsPanel(): JSX.Element {
...
+                  .then(() => {
+                    handleRemovePanelQueryParams();
+                    return;
+                  })
+                  .catch((error) => {
+                    setInlineErrors((prevInlineError) => [
```
Comment (huwshimi) [inline src/panels/ActionsPanel/ActionsPanel.tsx:256]: If you set the saving state when the action starts then you'll need to update it again here:

```suggestion
                  .catch((error) => {
                    setSaving(false);
                    setInlineErrors((prevInlineError) => [
```

File: src/panels/ActionsPanel/ActionsPanel.tsx [hunk 6]
```diff
@@ -119,9 +145,16 @@ export default function ActionsPanel(): JSX.Element {
             setActionData(actions.results[0].actions);
           }
           setFetchingActionData(false);
+          setInlineErrors((prevInlineErrors) => [null, prevInlineErrors[1]]);
           return;
         })
-        .catch((error) => console.error(Label.GET_ACTIONS_ERROR, error));
+        .catch((error) => {
+          setInlineErrors((prevInlineErrors) => [
```
Comment (huwshimi) [inline src/panels/ActionsPanel/ActionsPanel.tsx:152]: Here also we need to only update a single error slot so we don't accidentally remove errors if the number of errors in the tuple changes.

File: src/panels/ActionsPanel/ActionsPanel.tsx [hunk 7]
```diff
@@ -211,12 +244,21 @@ export default function ActionsPanel(): JSX.Element {
...
-                handleRemovePanelQueryParams();
+                executeAction()
+                  .then(() => {
+                    handleRemovePanelQueryParams();
+                    return;
+                  })
+                  .catch((error) => {
+                    setInlineErrors((prevInlineError) => [
+                      prevInlineError[0],
+                      Label.EXECUTE_ACTION_ERROR,
```
Comment (huwshimi) [inline src/panels/ActionsPanel/ActionsPanel.tsx:258]: Also need to only update a single slot in the tuple.

File: src/panels/ConfigPanel/ConfigPanel.tsx [hunk 8]
```diff
@@ -392,7 +395,7 @@ export default function ConfigPanel(): JSX.Element {
 
             <div className="config-panel__list">
               {formErrors
```
Comment (huwshimi) [inline src/panels/ConfigPanel/ConfigPanel.tsx:397]: I think this should also have `.some(...` or at the very least check that this isn't an empty array.

File: src/panels/ConfigPanel/ConfigPanel.tsx [hunk 9]
```diff
@@ -57,12 +57,15 @@ type ConfigQueryParams = {
...
-  ));
+const generateErrors = (errors: string[], scrollArea: HTMLElement | null) => (
+  <ScrollOnRender scrollArea={scrollArea}>
+    {errors.map((error) => (
+      <Notification key={error} severity="negative">
+        {error}
+      </Notification>
+    ))}
+  </ScrollOnRender>
+);
```
Comment (huwshimi) [inline src/panels/ConfigPanel/ConfigPanel.tsx:68]: Possibly in a follow-up branch this could become a shared component, something like `PanelErrors` and include the `.some(...)` like you do in the other panel, as we're doing the same thing in a number of places.

File: src/panels/ActionsPanel/ActionsPanel.tsx [hunk 10]
```diff
@@ -254,8 +296,12 @@ export default function ActionsPanel(): JSX.Element {
       data-testid={TestId.PANEL}
       title={generateTitle()}
       onRemovePanelQueryParams={handleRemovePanelQueryParams}
+      ref={scrollArea}
     >
       <p data-testid="actions-panel-unit-list">
+        {inlineErrors.some((error) => error)
+          ? generateInlineErrors(inlineErrors, scrollArea.current)
+          : null}
```
Comment (huwshimi) [inline src/panels/ActionsPanel/ActionsPanel.tsx:304]: This needs to be outside of the wrapping `<p>` as this is currently giving errors about nested `<p>` tags.

**Review by huwshimi — COMMENTED** [review]
I've ended up making quite a few small comments so I'll do another review of this once those changes have been made.

File: src/panels/ActionsPanel/ActionsPanel.tsx [hunk 11]
```diff
@@ -119,9 +128,21 @@ export default function ActionsPanel(): JSX.Element {
...
           setFetchingActionData(false);
+          setInlineErrors((prevInlineErrors) => {
+            const newInlineErrors = [...prevInlineErrors];
+            newInlineErrors[0] = null;
+            return newInlineErrors;
+          });
           return;
         })
-        .catch((error) => console.error(Label.GET_ACTIONS_ERROR, error));
+        .catch((error) => {
```
Comment (huwshimi) [inline src/panels/ActionsPanel/ActionsPanel.tsx:138]: I think this also needs to call `setFetchingActionData(false);` so that the spinner hides when the error appears.

File: src/panels/PanelInlineErrors/PanelInlineErrors.tsx [hunk 12]
```diff
@@ -0,0 +1,26 @@
+import { Notification } from "@canonical/react-components";
+
+import ScrollOnRender from "components/ScrollOnRender";
+
+type Props = {
+  inlineErrors: (string | null)[] | null;
+  scrollArea?: HTMLElement | null;
+};
+
+const PanelInlineErrors = ({
```
Comment (huwshimi) [inline src/panels/PanelInlineErrors/PanelInlineErrors.tsx:10]: Nice, thanks for adding this.

File: src/panels/PanelInlineErrors/PanelInlineErrors.test.tsx [hunk 13]
```diff
@@ -0,0 +1,11 @@
+import { render, screen } from "@testing-library/react";
+
+import PanelInlineErrors from "./PanelInlineErrors";
+
+describe("PanelInlineErrors", () => {
```
Comment (huwshimi) [inline src/panels/PanelInlineErrors/PanelInlineErrors.test.tsx:5]: I think it would be good to also test that the `ScrollOnRender` component does not appear if `inlineErrors` is `null`, an empty array or only contains falsey items.

**Review by huwshimi — APPROVED** [review]
Thanks for all the changes. I've made a couple of small comments but this can land once those have been fixed.

File: src/panels/ActionsPanel/ActionsPanel.tsx [hunk 14]
```diff
@@ -242,26 +290,51 @@ export default function ActionsPanel(): JSX.Element {
...
       data-testid={TestId.PANEL}
       title={generateTitle()}
       onRemovePanelQueryParams={handleRemovePanelQueryParams}
+      ref={scrollArea}
     >
+      <PanelInlineErrors
+        inlineErrors={
+          inlineErrors[0]
+            ? inlineErrors.map((error, index) =>
+                index === 0 ? (
```
Comment (huwshimi) [inline src/panels/ActionsPanel/ActionsPanel.tsx:312]: I think it might be nice to leave a comment here to explain why this is necessary.

**Review by huwshimi — APPROVED** [review]
Just one tiny comment.
